### DISCUSSION TITLE
Index building not using all the CPUs

### DISCUSSION THREAD

**Discussion by hadim** [milvus-io/milvus#5534] [Q&A and General discussion] [answered]
Apologize if such a question has been answered already but I could not find the answer.

---

Here is the setup: 

- 2B vectors, binary, size=512, metric=Jaccard, index=IVFLAT, nlist=65_536, index_file_size=4096
- Latest docker Milvus image (1.1.0)
- Milvus is running on a single instance (64CPUs and 500GB of memory).

I have three questions:

- 1) when using log level `info`, I could actually see the `index progress 2/55` logs. But the returned collection information isn't showing any `IVFLAT` segments but only `IDMAP` ones. Then I enabled log level `debug` and discovered that the `index progress 2/55` logs were only about creating the index file on disk and that the index building was happening after. Can you confirm this is correct?

- 2) The index being build uses all the 64 CPUs but only one is used at 100% and all the other ones are used at 20% only. Is there is a reason for that? Are multiple segments being indexed at the same time? At what level the concurrency happens, within a segment or in between all the segments? Do you know how I could make Milvus uses all the cores? Given the 20% usage, I guess using a bigger machine (200 CPUs) is not going to make the index building faster right?

- 3) Is there is a way to get a more granular progress status of the index building? As of now, I have no idea whether it's going to take a few hours, ~30h or a few days.

Thank you.

**Comment by yhmo** [comment, marked as answer]
- i. The info log "index progress x/xx" shows how many segments finished index. "index progress 2/55"  means there are 55 segments need to be indexed, 2 of them have finished.
Once create_index() is called, the info log will print "CreateIndex begin.", and after each segment finished index, the info log will print "Build Index Progress: x of x". After all segments finished index, the info log will print "CreateIndex end."
An example:
```
[2021-06-02 15:22:15,397][INFO][SERVER][CreateIndex][grpcpp_sync_ser] Request [5004] CreateIndex begin.
[2021-06-02 15:23:31,353][INFO][ENGINE][BackgroundBuildIndex][milvus_server] Build Index Progress: 1 of 4
[2021-06-02 15:24:31,570][INFO][ENGINE][BackgroundBuildIndex][milvus_server] Build Index Progress: 2 of 4
[2021-06-02 15:25:09,676][INFO][ENGINE][BackgroundBuildIndex][milvus_server] Build Index Progress: 3 of 4
[2021-06-02 15:26:11,626][INFO][ENGINE][BackgroundBuildIndex][milvus_server] Build Index Progress: 4 of 4
[2021-06-02 15:26:11,627][INFO][SERVER][CreateIndex][grpcpp_sync_ser] Request [5004] CreateIndex end.
```

- ii. For binary vector, we also noticed that not all cpu are busy during building index. We read the source code of FAISS, looks like there is space to improve the cpu usage. I just create an issue #5543 for this.

- iii. There is an issue #3690 about status of build index, we didn't find a good solution to implement it. It require insert some kind of "hook" into the FAISS level, will make maintenance difficult.

**Comment by hadim** [comment]
Thanks, @yhmo! 

According to CPU usage the index is currently being built. Given the size of the dataset (~3B of vectors) the number of vectors per segment is 60M. I set `nlist` to `60_000`. Do you have a sense of how much `nlist` impacts the index building time? Is a `nlist` of `60_000` ok for 50 segments of that size?

Looking at the database all the 50 segments are of `engine_type` `9` and `file_type` `2` (I guess it means `IDMAP`?) except for one segment that is `engine_type` `10` and `file_type` `6` (it's also the only one with a row count of `0`). Can you confirm this is the index being built? Also, is there is any reason multiple index segments can't be built in parallel?

**Reply by hadim** [reply]
After restarting the instances a couple of time, dropping the index and starting with a smaller `nlist` of `1000`, the index started to build much faster.

**Reply by yhmo** [reply]
Milvus merge segment size to be larger than index_file_size, not exactly equal to index_file_size, that means most of segments size is a bit larger that index_file_size. 
You set index_file_size=4096MB, so most of segments size is 4096+, may be 4100 or 4150. Except the "tail segment", its size is less than index_file_size. So for most of segments, vector count is 60M+.
This [page](https://milvus.io/docs/v1.1.0/performance_faq.md#how-to-set-nlist-and-nprobe-for-ivf-indexes) describe how to estimate nlist for IVF index. The recommended value of nlist is `4 × sqrt(n)`. For n=60M+, recommended nlist=31000. Large value of nlist makes build index very slow, so you see nlist=1000 is much faster than nlist=60000.

The engine_type definition:
```
enum class EngineType {
    INVALID = 0,
    FAISS_IDMAP = 1,
    FAISS_IVFFLAT = 2,
    FAISS_IVFSQ8 = 3,
    NSG_MIX = 4,
    FAISS_IVFSQ8H = 5,
    FAISS_PQ = 6,
    SPTAG_KDT = 7,
    SPTAG_BKT = 8,
    FAISS_BIN_IDMAP = 9,
    FAISS_BIN_IVFFLAT = 10,
    HNSW = 11,
    ANNOY = 12,
    MAX_VALUE = ANNOY,
};
```
For binary vector, engine_type=9 is IDMAP, 10 is IVFFLAT. 
engine_type is a "target index type" for segments, this value is come from its owner collection. You may see some item is raw file, but their engine_type is not IDMAP, because their target index type has been specified according to their owner collection.


The file_type definition:
```
typedef enum {
        NEW = 0,
        RAW = 1,
        TO_INDEX = 2,
        INDEX = 3,
        TO_DELETE = 4,
        NEW_MERGE = 5,
        NEW_INDEX = 6,
        BACKUP = 7,
    } FILE_TYPE;
```
file_type=2 means the raw file is ready to prepare for build index, once you invoke create_index(), all raw file whose size is larger than index_file_size will be marked as file_type=2(from 1) and put to the queue waiting to build index.
file_type=6 means this is an index file that in building, this is a placeholder for current on-going build index task, only has one such item in metadata. Once the task finished, this item will become file_type=3, and the original raw file will be marked as file_type=7

Milvus build index for each segment one by one, because build index is a intensive computing work, typically it occupy all cpu cores and cpu cache, if we build multiple index in parallel, they will waste time in competition for CPU.

**Reply by hadim** [reply]
Thanks for the details! Those types are very helpful.

**Reply by hadim** [reply]
Having access to those data via the API without calling queries to the DB would be helpful I think.
